api/layer/multipart_upload.go

```diff
@@ -0,0 +1,494 @@
+				zap.Stringer("bucket id", p.Bkt.CID),
+				zap.Error(err))
+			continue
+		}
+		info := uploadInfoFromMeta(meta, p.Prefix, p.Delimiter)
+		if info != nil {
+			uploads = append(uploads, info)
+		}
+	}
+
+	sort.Slice(uploads, func(i, j int) bool {
+		if uploads[i].Key == uploads[j].Key {
+			return uploads[i].UploadID < uploads[j].UploadID
+		}
+		return uploads[i].Key < uploads[j].Key
+	})
+
+	if p.KeyMarker != "" {
+		if p.UploadIdMarker != "" {
+			uploads = trimAfterUploadIDAndKey(p.KeyMarker, p.UploadIdMarker, uploads)
+		}
+		uploads = trimAfterUploadKey(p.KeyMarker, uploads)
+	}
+
+	if p.MaxUploads != 0 && len(uploads) > p.MaxUploads {
+		result.IsTruncated = true
+		uploads = uploads[:p.MaxUploads]
+		result.NextUploadIdMarker = uploads[len(uploads)-1].UploadID
+		result.NextKeyMarker = uploads[len(uploads)-1].Key
+	}
+
+	for _, ov := range uploads {
+		if ov.IsDir {
+			result.Prefixes = append(result.Prefixes, ov.Key)
+		} else {
+			result.Uploads = append(result.Uploads, ov)
+		}
+	}
+
+	return &result, nil
+}
+
+func (n *layer) AbortMultipartUpload(ctx context.Context, p *UploadInfoParams) error {
+	objects, err := n.getUploadParts(ctx, p.UploadID, p.Key, p.Bkt)
+	if err != nil {
+		return err
+	}
+
+	for _, info := range objects {
+		err := n.objectDelete(ctx, info.CID, info.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (n *layer) ListParts(ctx context.Context, p *ListPartsParams) (*ListPartsInfo, error) {
+	var res ListPartsInfo
+	objs, err := n.getUploadParts(ctx, p.Info.UploadID, p.Info.Key, p.Info.Bkt)
+	if err != nil {
+		return nil, err
+	}
+
+	res.Owner = objs[0].Owner
+
+	parts := make([]*Part, 0, len(objs))
+
+	for num, objInfo := range objs {
+		if num == 0 {
+			continue
+		}
+		parts = append(parts, &Part{
+			ETag:         objInfo.HashSum,
+			LastModified: objInfo.Created.Format(time.RFC3339),
+			PartNumber:   num,
+			Size:         objInfo.Size,
+		})
+	}
+
+	sort.Slice(parts, func(i, j int) bool {
+		return parts[i].PartNumber < parts[j].PartNumber
+	})
+
+	if p.PartNumberMarker != 0 {
+		for i, part := range parts {
+			if part.PartNumber > p.PartNumberMarker {
+				parts = parts[i:]
+				break
+			}
+		}
+		return &res, nil
+	}
+
+	if len(parts) > p.MaxParts {
+		res.IsTruncated = true
+		res.NextPartNumberMarker = parts[p.MaxParts-1].PartNumber
+		parts = parts[:p.MaxParts]
+	}
+
+	res.Parts = parts
+
+	return &res, nil
+}
+
+func (n *layer) GetUploadInitInfo(ctx context.Context, p *UploadInfoParams) (*data.ObjectInfo, error) {
+	filters := []*filter{
+		{attr: UploadIDAttributeName, val: p.UploadID},
+		{attr: UploadPartNumberAttributeName, val: "0"},
+		{attr: UploadKeyAttributeName, val: p.Key},
+	}
+	f := &findParams{
+		filters: filters,
+		cid:     p.Bkt.CID,
+	}
+	ids, err := n.objectSearch(ctx, f)
+	if err != nil {
+		if errors.IsS3Error(err, errors.ErrNoSuchKey) {
+			return nil, errors.GetAPIError(errors.ErrNoSuchUpload)
+		}
+		return nil, err
+	}
+	if ids == nil {
```

Comment:
It will be safer this way:
```suggestion
	if len(ids) == 0 {
```